File: py/pytanga/viz/templates/views/stack-view.js
// Tanga Viewer — `StackView`: a flow/flex container that stacks children
// vertically, horizontally, or wraps.  Content-size math lives in `stack-size.js`.

import { View } from './view.js';
import { GAP, stackMainAxis, stackMinSize, stackPreferredSize } from './stack-size.js';

const DIRECTIONS = ['vertical', 'horizontal', 'wrap'];

/**
 * A flex container of `View` children laid out in normal flow (no splitters).
 * `direction` is `"vertical"` (column), `"horizontal"` (row), or `"wrap"`
 * (row that wraps to a new line when out of space).
 */
export class StackView extends View {
    constructor({ direction = 'vertical', scrollable = false, children = [] } = {}) {
        super();
        if (!DIRECTIONS.includes(direction)) {
            throw new Error(
                `direction must be one of ${DIRECTIONS.join(', ')}, got ${JSON.stringify(direction)}`
            );
        }
        this.direction = direction;
        this.scrollable = scrollable;
        this.children = [];
        this._childSubs = new Map(); // view -> AbortController
        this._content = this.el; // children mount here (GroupView retargets this)

        this.el.classList.add('tanga-stack', `tanga-stack-${direction}`);
        this._applyFlex();
        this._applyScroll();

        for (const child of children) this.addChild(child);
    }

    _applyFlex() {
        const s = this._content.style;
        s.display = 'flex';
        s.position = 'relative';
        s.gap = `${GAP}px`;
        s.flexDirection = this.direction === 'vertical' ? 'column' : 'row';
        s.flexWrap = this.direction === 'wrap' ? 'wrap' : 'nowrap';
    }

    _applyScroll() {
        if (!this.scrollable) return;
        this._content.classList.add('tanga-scroll');
        Object.assign(this._content.style, {
            overflow: 'auto',
            minWidth: '0',
            minHeight: '0',
            flex: '1 1 auto',
        });
    }

    addChild(view) {
        const ac = new AbortController();
        view.mount(this._content);
        view.on('preferredchange', () => this._relayout(), { signal: ac.signal });
        view.on('constraintschange', () => this._relayout(), { signal: ac.signal });
        this._childSubs.set(view, ac);
        this.children.push(view);
        this._relayout();
        return view;
    }

    removeChild(view) {
        const idx = this.children.indexOf(view);
        if (idx === -1) return;
        const ac = this._childSubs.get(view);
        if (ac) ac.abort();
        this._childSubs.delete(view);
        this.children.splice(idx, 1);
        view.unmount();
        this._relayout();
    }

    // A stack's minimum is at least what its children need (content sizing).
    // A scrollable stack instead decouples from content along its main axis so
    // an enclosing SplitView may shrink it and the content scrolls.
    minSizePx(axis, available) {
        const explicit = super.minSizePx(axis, available);
        if (this.scrollable && axis === stackMainAxis(this.direction)) {
            return explicit;
        }
        return Math.max(explicit, stackMinSize(axis, this.direction, this.children, available));
    }

    preferredPx(axis, available) {
        const explicit = super.preferredPx(axis, available);
        if (explicit !== null && explicit !== undefined) return explicit;
        if (this.scrollable && axis === stackMainAxis(this.direction)) {
            return null;
        }
        return stackPreferredSize(axis, this.direction, this.children, available);
    }

    _relayout() {
        // Content size may have changed — notify an enclosing SplitView.
        this.emit('preferredchange', { fields: ['preferredWidth', 'preferredHeight'] });
    }

    destroy() {
        for (const ac of this._childSubs.values()) ac.abort();
        this._childSubs.clear();
        super.destroy();
    }
}

File: py/pytanga/viz/views/stack_view.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2021 Christian Perwass

"""The :class:`StackView` flow container."""

from __future__ import annotations

from typing import Any

from ._base import View
from ._enums import EStackAlign, EStackDirection, EStackJustify
from .._size import SizeSpec


class StackView(View):
    """A flow container that stacks children vertically, horizontally, or wraps.

    Unlike :class:`SplitView`, children flow in normal document order (no
    splitters) and the container sizes to its content along the stack axis.
    """

    _node_type = "stack"

    def __init__(
        self,
        direction: EStackDirection | str,
        children: list[View] | None = None,
        *,
        scrollable: bool = False,
        gap: int | None = None,
        align: EStackAlign | str = EStackAlign.STRETCH,
        justify: EStackJustify | str = EStackJustify.START,
        size: SizeSpec = None,
        preferred_width: SizeSpec = None,
        preferred_height: SizeSpec = None,
        min_width: SizeSpec = None,
        min_height: SizeSpec = None,
        max_width: SizeSpec = None,
        max_height: SizeSpec = None,
    ) -> None:
        super().__init__(
            size=size,
            preferred_width=preferred_width,
            preferred_height=preferred_height,
            min_width=min_width,
            min_height=min_height,
            max_width=max_width,
            max_height=max_height,
        )
        if direction not in EStackDirection:
            raise ValueError(
                f"direction must be 'vertical', 'horizontal' or 'wrap', got {direction!r}"
            )
        if align not in EStackAlign:
            raise ValueError(
                f"align must be 'start', 'center', 'end' or 'stretch', got {align!r}"
            )
        if justify not in EStackJustify:
            raise ValueError(
                f"justify must be one of 'start', 'center', 'end', 'space-between', "
                f"'space-around', 'space-evenly', got {justify!r}"
            )
        if gap is not None and (
            isinstance(gap, bool) or not isinstance(gap, int) or gap < 0
        ):
            raise ValueError(f"gap must be a non-negative int or None, got {gap!r}")
        self.direction = direction
        self.children = list(children or [])
        self.scrollable = scrollable
        self.gap = gap
        self.align = align
        self.justify = justify

    def _serialize(self) -> dict[str, Any]:
        result = super()._serialize()
        result["direction"] = self.direction
        result["scrollable"] = self.scrollable
        result["gap"] = self.gap
        result["align"] = self.align
        result["justify"] = self.justify
        result["children"] = [child._serialize() for child in self.children]
        return result
